### ISSUE TITLE
Uploads of big files stop at 50,0% file size

### ISSUE THREAD

**Issue by wdenx** [yadayada/acd_cli#253]
I'm observing a very strange phenomenon:  I'm trying to upload a bung
of files, some of which are in the 4.4 ... 6.7 GB range.

I've uploaded such big files in the past before, but now the upload is
failing reliably for all of them.  The strange thing is that the
upload always stops at exactly 50.05 of the fie size - no matter what
the actual file size is:

For a 4.4 GB file I would see for example:

[############             ]  50.0% of    4GiB  0/1     0.0B/s       0s

Return code: 0

In some cases I would see an error message like this:

(another upload of the same 4.4 GB file):

16-01-22 23:34:43.842 [ERROR] [acd_cli] - Size mismatch between local and remote file for "eldk-5.8-armv4t.iso".
[############             ]  50.0% of    4GiB  0/1     0.0B/s       0s
1 file(s) failed.

Return code: 128

The information in the log file does not reveal any problems to me:

16-01-22 23:45:05.646 [INFO] [acd_cli] - Uploading eldk-5.8-armv4t.iso
16-01-22 23:45:05.694 [DEBUG] [acdcli.api.backoff_req] - Retry 0, waiting -0.375748s
16-01-22 23:45:05.695 [INFO] [acdcli.api.backoff_req] - POST "https://content-na.drive.amazonaws.com/cdproxy/nodes"
16-01-22 23:45:05.695 [DEBUG] [acdcli.api.backoff_req] - <MultipartEncoder: OrderedDict([('metadata', '{"name": "eldk-5.8-armv4t.iso", "kind": "FILE", "parents": ["5acQL-AeQa2inrRcgnAk3w"]}'), ('content', ('eldk-5.8-armv4t.iso', <acdcli.api.content._TeeBufferedReader object at 0x7f1909850f28>, 'application/octet-stream'))])>
16-01-22 23:45:05.713 [INFO] [requests.packages.urllib3.connectionpool] - Starting new HTTPS connection (1): content-na.drive.amazonaws.com
16-01-23 00:16:43.695 [DEBUG] [requests.packages.urllib3.connectionpool] - "POST /cdproxy/nodes?suppress=deduplication HTTP/1.1" 201 532
16-01-23 00:16:43.850 [INFO] [acdcli.cache.sync] - Inserted/updated 1 file(s).
16-01-23 00:16:43.986 [INFO] [acdcli.cache.sync] - Parented 1 node(s).
16-01-23 00:16:44.017 [DEBUG] [acd_cli] - Local and remote hashes match for "eldk-5.8-armv4t.iso".

So I wonder what might be broken...

I speculate if the problem might be caused by a recent update to
Fedora 23, which means I am now using python3-3.4.3-5.fc23.x86_64

Has anybody seen this before?  Any ideas what to look for?

Thanks in advance,

Wolfgang

**Comment by wdenx** [comment]
Here is another example for a somewhat smaller (2.9G) file:

16-01-23 13:47:26.230 [ERROR] [acd_cli] - Size mismatch between local
and remote file for "eldk-5.8-mips.iso".
[############             ]  50.0% of    3GiB  0/1     0.0B/s       0s
1 file(s) failed.
=== RC=128

acd_cli.log saus:

16-01-23 13:26:34.120 [INFO] [acd_cli] - Plugin leaf classes: TestPlugin
16-01-23 13:26:34.120 [INFO] [acd_cli] - TestPlugin attached.
16-01-23 13:26:34.121 [INFO] [acdcli.api.client] - Initializing ACD with path "/home/wd/.cache/acd_cli".
16-01-23 13:26:34.121 [INFO] [acdcli.api.oauth] - AppspotOAuthHandler initialized
16-01-23 13:26:34.123 [INFO] [acdcli.cache.schema] - DB schema version is 2.
16-01-23 13:26:34.282 [INFO] [acd_cli] - Last sync at 2016-01-23 12:26:33.244750.
16-01-23 13:26:34.284 [INFO] [acd_cli] - Resolved "/public/eldk/5.8/iso" to "Node('5acQL-AeQa2inrRcgnAk3w', 'iso')"
16-01-23 13:26:34.284 [DEBUG] [acd_cli] - Namespace(acd_client=<acdcli.api.client.ACDClient object at 0x7f678c704e80>, action='ul', cache=<acdcli.cache.db.NodeCache object at 0x7f678c70fac8>, check=0, color=2, debug=None, deduplicate=False, exclude_fe=[], exclude_path=[], exclude_re=[], force=False, func=<function upload_action at 0x7f678c6fe8c8>, log=True, max_connections=1, max_retries=0, no_wait=False, overwrite=False, parent='5acQL-AeQa2inrRcgnAk3w', parent_path='/public/eldk/5.8/iso', path=['eldk-5.8-mips.iso'], remove_source_files=False, utf=False, verbose=None)
16-01-23 13:26:34.285 [INFO] [acdcli.utils.threading] - 1 jobs in queue.
16-01-23 13:26:34.296 [INFO] [acd_cli] - Uploading eldk-5.8-mips.iso
16-01-23 13:26:34.385 [DEBUG] [acdcli.api.backoff_req] - Retry 0, waiting -0.262511s
16-01-23 13:26:34.385 [INFO] [acdcli.api.backoff_req] - POST "https://content-na.drive.amazonaws.com/cdproxy/nodes"
16-01-23 13:26:34.386 [DEBUG] [acdcli.api.backoff_req] - <MultipartEncoder: OrderedDict([('metadata', '{"kind": "FILE", "parents": ["5acQL-AeQa2inrRcgnAk3w"], "name": "eldk-5.8-mips.iso"}'), ('content', ('eldk-5.8-mips.iso', <acdcli.api.content._TeeBufferedReader object at 0x7f678c6aa7f0>, 'application/octet-stream'))])>
16-01-23 13:26:34.403 [INFO] [requests.packages.urllib3.connectionpool] - Starting new HTTPS connection (1): content-na.drive.amazonaws.com
16-01-23 13:47:25.828 [DEBUG] [requests.packages.urllib3.connectionpool] - "POST /cdproxy/nodes?suppress=deduplication HTTP/1.1" 201 530
16-01-23 13:47:26.053 [INFO] [acdcli.cache.sync] - Inserted/updated 1 file(s).
16-01-23 13:47:26.228 [INFO] [acdcli.cache.sync] - Parented 1 node(s).
16-01-23 13:47:26.229 [DEBUG] [acd_cli] - Local and remote hashes match for "eldk-5.8-mips.iso".
16-01-23 13:47:26.230 [ERROR] [acd_cli] - Size mismatch between local and remote file for "eldk-5.8-mips.iso".


Mind the "Local and remote hashes match" message first, followed by
the "Size mismatch' error.

**Comment by yadayada** [comment]
I think this is (mainly) a duplicate of #234. And the local/remote size check isn't triggered in some cases; maybe @zmwangx will know what went wrong there.

Could you 
1) tell me the output of
```
pip3 freeze | grep requests
```
2) upgrade requests-toolbelt
3) try again
4) if necessary, upgrade requests

**Comment by wdenx** [comment]
re 1):
-> pip3 freeze | grep requests
You are using pip version 7.1.0, however version 8.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
requests==2.9.1
requests-file==1.4
requests-ftp==0.3.1
requests-toolbelt==0.5.1

re 2) I think requests-toolbelt is already at the latest version.  if not - where can I find a more recent one?

re 4) I think requests is already at the latest version.  if not - where can I find a more recent one?

**Comment by yadayada** [comment]
Upgrades are done with `pip3 install --upgrade <PACKAGE>`. But, as you say, I think you've got the newest ones. If you want to downgrade, the combination `pip3 install requests==2.8.1 requests-toolbelt==0.4.0` also works for me.

Maybe the commit above will help. That removes an old bundled module that, when used, I could reproduce this error with.

I'd also like to see if these local packages that pip shows are loaded. The output of
```
python3 -c "import requests, requests_toolbelt; print(requests); print(requests_toolbelt)" 
```
should include the same paths as the installation locations in
```
pip3 show requests
pip3 show requests-toolbelt
```

**Comment by wdenx** [comment]
-> python3 -c "import requests, requests_toolbelt; print(requests); print(requests_toolbelt)" 
<module 'requests' from '/usr/lib/python3.4/site-packages/requests/__init__.py'>
<module 'requests_toolbelt' from '/usr/lib/python3.4/site-packages/requests_toolbelt/__init__.py'>
-> pip3 show requests
---
Metadata-Version: 1.1
Name: requests
Version: 2.9.1
Summary: Python HTTP for Humans.
Home-page: http://python-requests.org
Author: Kenneth Reitz
Author-email: me@kennethreitz.com
License: Apache 2.0
Location: /usr/lib/python3.4/site-packages
Requires: urllib3
-> pip3 show requests-toolbelt
---
Metadata-Version: 2.0
Name: requests-toolbelt
Version: 0.4.0
Summary: A utility belt for advanced users of python-requests
Home-page: https://toolbelt.readthedocs.org
Author: Ian Cordasco, Cory Benfield
Author-email: graffatcolmingov@gmail.com
License: Copyright 2014 Ian Cordasco, Cory Benfield
Location: /usr/lib/python3.4/site-packages
Requires: requests
-> 

Wow... your latest commit has fixed the problem for me; 3 attempts = 3 successful uploads so far (4.4G each).

Thanks a lot!

**Comment by zmwangx** [comment]
The problem has been solved, but regarding the claim

> And the local/remote size check isn't triggered in some cases; maybe @zmwangx will know what went wrong there.

I can't think of a case when the size check isn't triggered (I consider it a bug if there is one). Both of OP's examples fail with 128, that is, `SIZE_MISMATCH`, so #236 is doing its job. Any counterexample?

**Comment by yadayada** [comment]
>Any counterexample?

I was thinking about his first example.

>For a 4.4 GB file I would see for example:
>
>[############ ] 50.0% of 4GiB 0/1 0.0B/s 0s
>
>Return code: 0

But in hindsight I think that might have been an older version without the size check. Never mind.

**Comment by zmwangx** [comment]
> I was thinking about his first example.

Sorry, I was reading on my phone and somehow missed that example... Yeah, I would guess that OP updated acd_cli between example 1 and example 2. @wdenx Could you please confirm?

**Comment by wdenx** [comment]
Confirmed, but it was the other way round: I downgraded to an older version which was working fine for me before.  Sorry for not clearly mentioning which test was run with which version.

**Comment by zmwangx** [comment]
Wait, I'm confused. The newer version exits with 0, then you downgraded to an older version which exits with 128? But the size check was added in 7e406de2def5324680f7168acd37d17980ea0dc5 (#236), which had not been merged until eight days ado.

**Comment by wdenx** [comment]
The rc=0 was with an older version:
e584e09   2015-11-10 17:09:30 +0100   add --size-bytes option to ls -l action

**Comment by zmwangx** [comment]
Okay, that clears it up.

**Comment by yadayada** [comment]
Hmm, I saw that Wolfgang also downgraded to requests-toolbelt 0.4.0...

Fiddling around with different requests and requests-toolbelt versions, it seems for me that only requests-toolbelt 0.5.0 will cause incomplete uploads independent of the requests version. Can anyone confirm/falsify this?

**Comment by zmwangx** [comment]
I've uploaded hundreds of big files since skipping 0.5.0, and haven't had a single problem (except for uploading too fast and getting timeout error after upload finishes).
